File: openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/Box4.java
package io.lacuna.artifex;

import static io.lacuna.artifex.Vec.vec;
import static java.lang.Double.NaN;

/**
 * @author ztellman
 */
public class Box4 extends Box<Vec4, Box4> {

  public static final Box4 EMPTY = new Box4(vec(NaN, NaN, NaN, NaN), vec(NaN, NaN, NaN, NaN));

  private final double lx, ly, lz, lw, ux, uy, uz, uw;

  private Box4(double ax, double ay, double az, double aw, double bx, double by, double bz, double bw) {
    if (ax < bx) {
      this.lx = ax;
      this.ux = bx;
    } else {
      this.ux = ax;
      this.lx = bx;
    }

    if (ay < by) {
      this.ly = ay;
      this.uy = by;
    } else {
      this.uy = ay;
      this.ly = by;
    }

    if (az < bz) {
      this.lz = az;
      this.uz = bz;
    } else {
      this.uz = az;
      this.lz = bz;
    }

    if (aw < bw) {
      this.lw = aw;
      this.uw = bw;
    } else {
      this.uw = aw;
      this.lw = bw;
    }
  }

  public Box4(Vec4 a, Vec4 b) {
    this(a.x, a.y, a.z, a.w, b.x, b.y, b.z, b.w);
  }

  public Box3 box3() {
    return new Box3(lx, ly, lz, ux, uy, uz);
  }

  public Box2 box2() {
    return new Box2(lx, ly, ux, uy);
  }

  @Override
  protected Box4 construct(Vec4 lower, Vec4 upper) {
    return new Box4(lower, upper);
  }

  @Override
  protected Box4 empty() {
    return EMPTY;
  }

  @Override
  public Vec4 lower() {
    return new Vec4(lx, ly, lz, lw);
  }

  @Override
  public Vec4 upper() {
    return new Vec4(ux, uy, uz, uw);
  }

  @Override
  public boolean isEmpty() {
    return this == EMPTY;
  }
}

File: openrndr-kartifex/src/commonMain/kotlin/Box4.kt
package org.openrndr.kartifex

/**
 * Represents a four-dimensional bounding box defined by two Vec4 points.
 * This class extends the abstract class `Box`, which provides common functionality for multi-dimensional bounding boxes.
 * The box is defined by its lower and upper bounds in 4D space.
 *
 * This class provides additional capabilities for extracting lower-dimensional bounding boxes (e.g., 3D and 2D boxes),
 * creating empty boxes, and determining spatial relationships such as containment and intersection.
 *
 * @constructor Creates a `Box4` defined by the given lower and upper bounds in all four dimensions.
 * @param ax The x-coordinate of the first point.
 * @param ay The y-coordinate of the first point.
 * @param az The z-coordinate of the first point.
 * @param aw The w-coordinate of the first point.
 * @param bx The x-coordinate of the second point.
 * @param by The y-coordinate of the second point.
 * @param bz The z-coordinate of the second point.
 * @param bw The w-coordinate of the second point.
 */
class Box4 private constructor(
    ax: Double,
    ay: Double,
    az: Double,
    aw: Double,
    bx: Double,
    by: Double,
    bz: Double,
    bw: Double
) :
    Box<Vec4, Box4>() {
    private val lx: Double
    private val ly: Double
    private val lz: Double
    private val lw: Double
    private val ux: Double
    private val uy: Double
    private val uz: Double
    private val uw: Double

    constructor(a: Vec4, b: Vec4) : this(a.x, a.y, a.z, a.w, b.x, b.y, b.z, b.w)

    fun box2(): Box2 {
        return Box2(lx, ly, ux, uy)
    }

    override fun construct(a: Vec4, b: Vec4): Box4 {
        return Box4(a, b)
    }

    override fun empty(): Box4 {
        return EMPTY
    }

    override fun lower(): Vec4 {
        return Vec4(lx, ly, lz, lw)
    }

    override fun upper(): Vec4 {
        return Vec4(ux, uy, uz, uw)
    }

    override val isEmpty: Boolean
        get() = this === EMPTY

    companion object {
        val EMPTY = Box4(
            Vec4(Double.NaN, Double.NaN, Double.NaN, Double.NaN), Vec4(
                Double.NaN, Double.NaN, Double.NaN, Double.NaN
            )
        )
    }

    init {
        if (ax < bx) {
            lx = ax
            ux = bx
        } else {
            ux = ax
            lx = bx
        }
        if (ay < by) {
            ly = ay
            uy = by
        } else {
            uy = ay
            ly = by
        }
        if (az < bz) {
            lz = az
            uz = bz
        } else {
            uz = az
            lz = bz
        }
        if (aw < bw) {
            lw = aw
            uw = bw
        } else {
            uw = aw
            lw = bw
        }
    }
}
